### DISCUSSION TITLE
Updating Comsol input and output

### DISCUSSION THREAD

**Discussion by Armin268** [MPh-py/MPh#136] [General]
Hi,

I want to update the COMSOL input and output in each iteration. The input is a dynamic text file, with data that varies in each iteration. On the other hand, I have asked COMSOL to generate a text file as output. However, when I use model.solve('Study 1') in each step, mph cannot update the input and output, and only solves the problem for the data from the first step. 

Is there any way to solve this issue using mph?

I am stuck on this, and any help would be appreciated.

**Comment by Armin268** [comment]
![Untitled](https://user-images.githubusercontent.com/122246737/233855624-a14bb674-aac5-4c8b-a09c-9473f400762e.png)

**Comment by john-hen** [comment]
I guess you'll have to figure out what commands to call that correspond to clicking that "Refresh" button in the Comsol GUI. See #86 for guidance.

Please note that MPh itself has very little to do with any of that. It's just a scripting interface. It does not solve the model, Comsol does. Calling `model.solve()` is like clicking "Compute" on the study node in the GUI. And if Comsol does not automatically reload your input file before it (Comsol) solves it, which the presence of the "Refresh" button suggests, then you'll have to tell Comsol that explicitly.

**Comment by Armin268** [comment]
Thank you very much for the answer.
As I got you, I should use Java for this automation. (I don't want to use MATLAB).
I extracted the Java code from my model I here is the part I was looking for:

    model.func("int1").set("source", "file");
    model.func("int1").set("funcs", new String[][]{{"por", "1"}});
    model.func("int1").set("filename", "C:\\Users\\armin\\Desktop\\DEM_CDF_2\\por.txt");
    model.func("int1").set("nargs", 3);
    model.func("int1").set("fununit", new String[]{""});
    model.func("int1").set("argunit", new String[]{"m", "m", "m"}); 
 
Now, can I use it based on something you explained  in "Access the full Comsol API" to force COMSOL to read the text file in each iteration?

Thanks,
Armin

**Reply by john-hen** [reply]
"Iteration" of what? What is iterating? Do you have a `for` loop or something in your code that calls `model.solve()` repeatedly? If so, add the code inside that loop and replace Java constructs such as `new String[]{""}` with the equivalent Python syntax, in that example just the empty string `""`. It needs to be Python code in the end, not Java. (Unless you actually want to use Java for the automation, then you don't need MPh or Python at all.)

**Reply by Armin268** [reply]
I appreciate your answer and your help.
Let me explain a part of my modelling. ( I don't want to take your time)
I coupled COMSOL (FEM-Based) with a 3rd Party software named PFC3D (DEM-Based) to model a solid-fluid interaction. 
To create this connection, I have to use Python (not Java or MATLAB) because PFC3D forced me to do that.
The "Iteration" in my case means one computational cycle between COMSOL and PFC3D. 
In each "Iteration", COMSOL should read some parameters from PFC3D (as a text file) and send velocity, pressure and gradient of pressure to PFC3D (as a text file =>('output1.txt')).
That is why I am trying to solve this automation process.
I defined a "while True:" to enforce both software to start solving the governing equation (in the following, I've shown one part of my code inside "while True:"

        #Solve the COMSOL model
        #=============================================
        client = mph.start(cores=1)
        model = client.load('Test2.mph')
        model.solve('Study 1')

        data = np.loadtxt('output1.txt')
        pressure = np.zeros(nele)
        gradp = np.zeros((nele,3))
        velocity = np.zeros((nele,3))

        # Assign values to the arrays

        velocity[:, 0] = data[:, 3]
        velocity[:, 1] = data[:, 4]
        velocity[:, 2] = data[:, 5]
        pressure[:] = data[:, 6]
        gradp[:, 0] = data[:, 7]
        gradp[:, 1] = data[:, 8]
        gradp[:, 2] = data[:, 9]
        
        #==== Send data to PFC3D ======

        pfc_link.send_data(pressure)
        pfc_link.send_data(gradp)
        pfc_link.send_data(velocity)

Is there any way to add Java code here to update my COMSOL model's input and output?
This is the last part of my modeling and I really need help.

Thank you in advance.

**Reply by john-hen** [reply]
First of all, given what you explained now, it looks like there is a way to do that. (I'm pretty sure. Though I cannot be a hundred percent sure.) And I'm glad this library is useful in that endeavor. Because coupling Comsol with another software that only has a Python interface... that's really hard to do with just the tools that Comsol provides. MPh can help there.

Now, what the exact code is that you need to write... hard for me to say. I can't run any of this to even test. All I can do is guess. Side note: This is why, in the tutorial, I point readers to an actual Comsol model file they can follow along with: `capacitor.mph`. Just so things are less abstract.

The Java code you coaxed out of Comsol should look "something like" this in Python:
```python
model.java.func("int1").set("source", "file")
model.java.func("int1").set("funcs", [["por", "1"]])
model.java.func("int1").set("filename", "C:\\Users\\armin\\Desktop\\DEM_CDF_2\\por.txt")
model.java.func("int1").set("nargs", 3)
model.java.func("int1").set("fununit", "")
model.java.func("int1").set("argunit", ["m", "m", "m"])
```

Not sure about the second line, the double list: `[["por", "1"]]`. Again, can't test this.

More importantly, that piece of code should go before `model.solve('Study')`.

For the record, I don't understand why you reload the model *inside* the loop (`model = client.load('Test2.mph')`), nor why you'd restart the client in every iteration (`client = mph.start(cores=1)`). That should not be necessary. And in fact, the client can't actually be restarted. I don't remember if that raises a warning (I changed that at some point), but it certainly achieves nothing.

Load the model once, before the loop, then update the interpolation function (is it?) in every iteration. Along the lines of the code above.

Also also, no need to use just one core: `cores=1`. Why not use all cores?

I don't see how you trigger the update of the output. There's nothing there that would update `output1.txt`. Comsol does not export data to files while it's solving, for all I know. So you would need code there that "exports" whatever data should go into that file after the solution was computed. Maybe you have an "Export" node for that in your model? Then that node needs to "run".

As mentioned in the discussion I linked to earlier, there are better, i.e. "more readable", ways, to write the code, like the snippet above. Using MPh's API, with something like `node = model/'functions'/'Interpolation1'` and then `node.property('source', 'file')` etc. But get it to run first. Don't test the whole thing. Consider writing a demo script that just loads the model, updates the input and exports the output. Once that works, the rest should be straightforward.

**Comment by Armin268** [comment]
I appreciate your help.
By doing some changes in my codes based on your advices, now it is working!!!!

THANK YOU.

**Reply by zhujinzhuo** [reply]
Hello, could you please share the complete code for coupling these two software?
Thank you
